File: android/src/main/kotlin/com/appsprint/flutter/AppSprintFlutterPlugin.kt
package com.appsprint.flutter

import android.content.Context
import com.appsprint.sdk.AppSprint
import com.appsprint.sdk.AppSprintConfig
import com.appsprint.sdk.AppSprintEventType
import com.appsprint.sdk.AppSprintNative
import com.appsprint.sdk.AttributionResult
import com.appsprint.sdk.GoogleAdsConsent
import com.appsprint.sdk.GoogleAdsConsentStatus
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class AppSprintFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private var bridgeExecutor: ExecutorService = newBridgeExecutor()

    private fun newBridgeExecutor(): ExecutorService = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "AppSprintFlutterBridge").apply { isDaemon = true }
    }

    private fun sdk(): AppSprint = AppSprint.shared(context)

    private fun runAsync(result: MethodChannel.Result, errorCode: String, block: () -> Unit) {
        bridgeExecutor.execute {
            try {
                block()
            } catch (e: Exception) {
                result.error(errorCode, e.message, null)
            }
        }
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        if (bridgeExecutor.isShutdown) {
            bridgeExecutor = newBridgeExecutor()
        }
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "appsprint_flutter/native")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {

            // Core SDK

            "configure" -> {
                val apiKey = (call.argument<String>("apiKey") ?: "").trim()
                if (apiKey.isEmpty()) {
                    result.error("CONFIGURE_ERROR", "AppSprint.configure requires a non-empty apiKey.", null)
                    return
                }
                runAsync(result, "CONFIGURE_ERROR") {
                    val config = AppSprintConfig(
                        apiKey = apiKey,
                        apiUrl = call.argument<String>("apiUrl")
                            ?: call.argument<String>("endpointBaseUrl")
                            ?: "https://api.appsprint.app",
                        enableAppleAdsAttribution = call.argument<Boolean>("enableAppleAdsAttribution") ?: true,
                        isDebug = call.argument<Boolean>("isDebug") ?: false,
                        logLevel = call.argument<Int>("logLevel") ?: if (call.argument<Boolean>("isDebug") == true) 0 else 2,
                        customerUserId = call.argument<String>("customerUserId"),
                        autoTrackSessions = call.argument<Boolean>("autoTrackSessions") ?: true,
                        autoRefreshAttribution = call.argument<Boolean>("autoRefreshAttribution") ?: true,
                        googleAdsConsent = googleAdsConsentFrom(call.argument<Map<String, Any?>>("googleAdsConsent")),
                    )
                    sdk().configure(config)
                    result.success(true)
                }
            }

            "sendEvent" -> {
                val eventTypeStr = call.argument<String>("eventType") ?: "custom"
                runAsync(result, "SEND_EVENT_ERROR") {
                    val type = AppSprintEventType.entries.find { it.wireValue == eventTypeStr } ?: AppSprintEventType.CUSTOM
                    val name = call.argument<String>("name")
                    val params = mutableMapOf<String, Any?>()
                    call.argument<Map<String, Any?>>("parameters")?.forEach { (k, v) -> params[k] = v }
                    val revenue = numberArgument(call, "revenue")
                    val currency = call.argument<String>("currency")
                    call.argument<Map<String, Any?>>("googleAdsConsent")?.let { params["googleAdsConsent"] = it }
                    if (revenue != null) params["revenue"] = revenue
                    if (currency != null) params["currency"] = currency
                    sdk().sendEvent(type, name, if (params.isNotEmpty()) params else null)
                    result.success(true)
                }
            }

            "sendTestEvent" -> {
                runAsync(result, "TEST_EVENT_ERROR") {
                    val r = sdk().sendTestEvent()
                    result.success(mapOf("success" to r.success, "message" to r.message))
                }
            }

            "flush" -> {
                runAsync(result, "FLUSH_ERROR") {
                    sdk().flush()
                    result.success(null)
                }
            }

            "clearData" -> {
                runAsync(result, "CLEAR_DATA_ERROR") {
                    sdk().clearData()
                    result.success(null)
                }
            }

            "setCustomerUserId" -> {
                val userId = call.argument<String>("userId") ?: ""
                runAsync(result, "SET_USER_ID_ERROR") {
                    sdk().setCustomerUserId(userId)
                    result.success(null)
                }
            }

            "refreshAttribution" -> {
                runAsync(result, "REFRESH_ATTRIBUTION_ERROR") {
                    result.success(sdk().refreshAttribution()?.let { attributionToMap(it) })
                }
            }

            "enableAppleAdsAttribution" -> {
                runAsync(result, "APPLE_ADS_ERROR") {
                    result.success(sdk().enableAppleAdsAttribution())
                }
            }

            "getAppSprintId" -> result.success(sdk().getAppSprintId())

            "getAttribution" -> {
                val attr = sdk().getAttribution()
                if (attr == null) {
                    result.success(null)
                    return
                }
                result.success(attributionToMap(attr))
            }

            "getAttributionParams" -> result.success(sdk().getAttributionParams())

            "isInitialized" -> result.success(sdk().isInitialized())

            "isSdkDisabled" -> result.success(sdk().isSdkDisabled())

            "destroy" -> {
                runAsync(result, "DESTROY_ERROR") {
                    sdk().destroy()
                    result.success(null)
                }
            }

            // Utility

            "getDeviceInfo" -> {
                runAsync(result, "DEVICE_INFO_ERROR") {
                    val deviceInfo = AppSprintNative(context).getDeviceInfo(includeAdvertisingId = true)
                    result.success(mapOf(
                        "deviceModel" to deviceInfo.deviceModel,
                        "screenWidth" to deviceInfo.screenWidth,
                        "screenHeight" to deviceInfo.screenHeight,
                        "nativeScreenWidth" to deviceInfo.nativeScreenWidth,
                        "nativeScreenHeight" to deviceInfo.nativeScreenHeight,
                        "screenScale" to deviceInfo.screenScale,
                        "hardwareConcurrency" to deviceInfo.hardwareConcurrency,
                        "processorCount" to deviceInfo.processorCount,
                        "maxTouchPoints" to deviceInfo.maxTouchPoints,
                        "memoryGb" to deviceInfo.memoryGb,
                        "lowPowerMode" to deviceInfo.lowPowerMode,
                        "batteryState" to deviceInfo.batteryState,
                        "batteryLevelBucket" to deviceInfo.batteryLevelBucket,
                        "preferredLanguages" to deviceInfo.preferredLanguages,
                        "timezoneOffsetMinutes" to deviceInfo.timezoneOffsetMinutes,
                        "deviceManufacturer" to deviceInfo.deviceManufacturer,
                        "deviceBrand" to deviceInfo.deviceBrand,
                        "deviceProduct" to deviceInfo.deviceProduct,
                        "deviceHardware" to deviceInfo.deviceHardware,
                        "gpuVendor" to deviceInfo.gpuVendor,
                        "gpuRenderer" to deviceInfo.gpuRenderer,
                        "connectionType" to deviceInfo.connectionType,
                        "networkType" to deviceInfo.networkType,
                        "carrierName" to deviceInfo.carrierName,
                        "carrierCountryCode" to deviceInfo.carrierCountryCode,
                        "mobileCountryCode" to deviceInfo.mobileCountryCode,
                        "mobileNetworkCode" to deviceInfo.mobileNetworkCode,
                        "sdkPlatform" to deviceInfo.sdkPlatform,
                        "sdkVersion" to deviceInfo.sdkVersion,
                        "locale" to deviceInfo.locale,
                        "timezone" to deviceInfo.timezone,
                        "osVersion" to deviceInfo.osVersion,
                        "appVersion" to deviceInfo.appVersion,
                        "gaid" to deviceInfo.gaid,
                    ).filterValues { it != null })
                }
            }

            "getAdServicesToken" -> result.success(null) // iOS only

            "requestTrackingAuthorization" -> result.success(AppSprintNative(context).requestTrackingAuthorization())

            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        bridgeExecutor.shutdown()
    }

    private fun numberArgument(call: MethodCall, key: String): Double? {
        val value = call.argument<Any>(key) ?: return null
        return when (value) {
            is Number -> value.toDouble().takeIf { it.isFinite() }
            is String -> value.trim().toDoubleOrNull()?.takeIf { it.isFinite() }
            else -> null
        }
    }

    private fun googleAdsConsentFrom(value: Map<String, Any?>?): GoogleAdsConsent? {
        val status = googleAdsConsentStatus(value?.get("adUserData")) ?: return null
        return GoogleAdsConsent(status)
    }

    private fun googleAdsConsentStatus(value: Any?): GoogleAdsConsentStatus? {
        val normalized = (value as? String)?.trim()?.uppercase(java.util.Locale.US) ?: return null
        return GoogleAdsConsentStatus.entries.firstOrNull {
            it.wireValue == normalized || it.name == normalized
        }
    }

    private fun attributionToMap(attr: AttributionResult): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "isAttributed" to attr.isAttributed,
            "source" to attr.source,
            "confidence" to attr.confidence,
        )
        attr.matchType?.let { map["matchType"] = it }
        attr.campaignName?.let { map["campaignName"] = it }
        attr.link?.let { map["link"] = mapOf("id" to it.id, "name" to it.name) }
        attr.appleAds?.let {
            map["appleAds"] = mutableMapOf<String, Any?>("campaignId" to it.campaignId).apply {
                it.orgId?.let { value -> put("orgId", value) }
                it.adGroupId?.let { value -> put("adGroupId", value) }
                it.keywordId?.let { value -> put("keywordId", value) }
                it.adId?.let { value -> put("adId", value) }
                it.countryOrRegion?.let { value -> put("countryOrRegion", value) }
                it.claimType?.let { value -> put("claimType", value) }
                it.clickDate?.let { value -> put("clickDate", value) }
                it.impressionDate?.let { value -> put("impressionDate", value) }
                it.conversionType?.let { value -> put("conversionType", value) }
                it.supplyPlacement?.let { value -> put("supplyPlacement", value) }
            }
        }
        attr.utmSource?.let { map["utmSource"] = it }
        attr.utmMedium?.let { map["utmMedium"] = it }
        attr.utmCampaign?.let { map["utmCampaign"] = it }
        attr.utmContent?.let { map["utmContent"] = it }
        attr.utmTerm?.let { map["utmTerm"] = it }
        return map
    }
}

File: ios/Classes/AppSprintFlutterPlugin.swift
import Flutter
import UIKit
import AppSprintSDK

public class AppSprintFlutterPlugin: NSObject, FlutterPlugin {

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "appsprint_flutter/native", binaryMessenger: registrar.messenger())
    let instance = AppSprintFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {

    // MARK: - Core SDK

    case "configure":
      guard let args = call.arguments as? [String: Any],
            let apiKey = args["apiKey"] as? String, !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        result(FlutterError(code: "CONFIGURE_ERROR", message: "AppSprint.configure requires a non-empty apiKey.", details: nil))
        return
      }
      Task { @MainActor in
        let enableAppleAds = args["enableAppleAdsAttribution"] as? Bool ?? true
        let isDebug = args["isDebug"] as? Bool ?? false
        let logLevelRaw = args["logLevel"] as? Int
        let customerUserId = args["customerUserId"] as? String
        let autoTrackSessions = args["autoTrackSessions"] as? Bool ?? true
        let autoRefreshAttribution = args["autoRefreshAttribution"] as? Bool ?? true
        let googleAdsConsent = Self.googleAdsConsent(from: args["googleAdsConsent"])

        let logLevel: AppSprintLogLevel
        if let raw = logLevelRaw, let level = AppSprintLogLevel(rawValue: raw) {
          logLevel = level
        } else {
          logLevel = isDebug ? .debug : .warn
        }

        var sdkConfig = AppSprintConfig(
          apiKey: apiKey,
          enableAppleAdsAttribution: enableAppleAds,
          isDebug: isDebug,
          logLevel: logLevel,
          customerUserId: customerUserId,
          autoTrackSessions: autoTrackSessions,
          autoRefreshAttribution: autoRefreshAttribution,
          googleAdsConsent: googleAdsConsent
        )

        let apiUrl = (args["apiUrl"] as? String) ?? (args["endpointBaseUrl"] as? String)
        if let urlString = apiUrl, let url = URL(string: urlString) {
          sdkConfig = AppSprintConfig(
            apiKey: apiKey,
            apiURL: url,
            enableAppleAdsAttribution: enableAppleAds,
            isDebug: isDebug,
            logLevel: logLevel,
            customerUserId: customerUserId,
            autoTrackSessions: autoTrackSessions,
            autoRefreshAttribution: autoRefreshAttribution,
            googleAdsConsent: googleAdsConsent
          )
        }

        await AppSprint.shared.configure(sdkConfig)
        result(true)
      }

    case "sendEvent":
      guard let args = call.arguments as? [String: Any],
            let eventTypeStr = args["eventType"] as? String else {
        result(FlutterError(code: "SEND_EVENT_ERROR", message: "eventType is required", details: nil))
        return
      }
      Task { @MainActor in
        let type = AppSprintEventType(rawValue: eventTypeStr) ?? .custom
        let name = args["name"] as? String
        var params: [String: Any]? = args["parameters"] as? [String: Any]

        if let rev = Self.numberValue(args["revenue"] ?? args["price"]) {
          if params == nil { params = [:] }
          params?["revenue"] = rev
        }
        if let cur = args["currency"] as? String {
          if params == nil { params = [:] }
          params?["currency"] = cur
        }
        if let googleAdsConsent = args["googleAdsConsent"] as? [String: Any] {
          if params == nil { params = [:] }
          params?["googleAdsConsent"] = googleAdsConsent
        }

        await AppSprint.shared.sendEvent(type, name: name, params: params)
        result(true)
      }

    case "sendTestEvent":
      Task { @MainActor in
        let r = await AppSprint.shared.sendTestEvent()
        result(["success": r.success, "message": r.message])
      }

    case "flush":
      Task { @MainActor in
        await AppSprint.shared.flush()
        result(nil)
      }

    case "clearData":
      Task { @MainActor in
        AppSprint.shared.clearData()
        result(nil)
      }

    case "setCustomerUserId":
      guard let args = call.arguments as? [String: Any],
            let userId = args["userId"] as? String else {
        result(FlutterError(code: "SET_USER_ID_ERROR", message: "userId is required", details: nil))
        return
      }
      Task { @MainActor in
        await AppSprint.shared.setCustomerUserId(userId)
        result(nil)
      }

    case "refreshAttribution":
      Task { @MainActor in
        guard let attr = await AppSprint.shared.refreshAttribution() else {
          result(nil)
          return
        }
        result(Self.attributionToDictionary(attr))
      }

    case "enableAppleAdsAttribution":
      Task { @MainActor in
        result(AppSprint.shared.enableAppleAdsAttribution())
      }

    case "getAppSprintId":
      Task { @MainActor in
        result(AppSprint.shared.getAppSprintId())
      }

    case "getAttribution":
      Task { @MainActor in
        guard let attr = AppSprint.shared.getAttribution() else {
          result(nil)
          return
        }
        result(Self.attributionToDictionary(attr))
      }

    case "getAttributionParams":
      Task { @MainActor in
        result(AppSprint.shared.getAttributionParams())
      }

    case "isInitialized":
      Task { @MainActor in
        result(AppSprint.shared.isInitialized)
      }

    case "isSdkDisabled":
      Task { @MainActor in
        result(AppSprint.shared.isSdkDisabled())
      }

    case "destroy":
      Task { @MainActor in
        AppSprint.shared.destroy()
        result(nil)
      }

    // MARK: - Utility

    case "getDeviceInfo":
      Task { @MainActor in
        let info = AppSprintNative.getDeviceInfo()
        var dict: [String: Any] = [:]
        if let m = info.deviceModel { dict["deviceModel"] = m }
        if let w = info.screenWidth { dict["screenWidth"] = w }
        if let h = info.screenHeight { dict["screenHeight"] = h }
        if let w = info.nativeScreenWidth { dict["nativeScreenWidth"] = w }
        if let h = info.nativeScreenHeight { dict["nativeScreenHeight"] = h }
        if let scale = info.screenScale { dict["screenScale"] = scale }
        if let concurrency = info.hardwareConcurrency { dict["hardwareConcurrency"] = concurrency }
        if let count = info.processorCount { dict["processorCount"] = count }
        if let touch = info.maxTouchPoints { dict["maxTouchPoints"] = touch }
        if let memory = info.memoryGb { dict["memoryGb"] = memory }
        if let lowPower = info.lowPowerMode { dict["lowPowerMode"] = lowPower }
        if let state = info.batteryState { dict["batteryState"] = state }
        if let bucket = info.batteryLevelBucket { dict["batteryLevelBucket"] = bucket }
        if let languages = info.preferredLanguages { dict["preferredLanguages"] = languages }
        if let offset = info.timezoneOffsetMinutes { dict["timezoneOffsetMinutes"] = offset }
        if let value = info.deviceManufacturer { dict["deviceManufacturer"] = value }
        if let value = info.deviceBrand { dict["deviceBrand"] = value }
        if let value = info.deviceProduct { dict["deviceProduct"] = value }
        if let value = info.deviceHardware { dict["deviceHardware"] = value }
        if let value = info.gpuVendor { dict["gpuVendor"] = value }
        if let value = info.gpuRenderer { dict["gpuRenderer"] = value }
        if let value = info.connectionType { dict["connectionType"] = value }
        if let value = info.networkType { dict["networkType"] = value }
        if let value = info.carrierName { dict["carrierName"] = value }
        if let value = info.carrierCountryCode { dict["carrierCountryCode"] = value }
        if let value = info.mobileCountryCode { dict["mobileCountryCode"] = value }
        if let value = info.mobileNetworkCode { dict["mobileNetworkCode"] = value }
        if let value = info.sdkPlatform { dict["sdkPlatform"] = value }
        if let value = info.sdkVersion { dict["sdkVersion"] = value }
        if let l = info.locale { dict["locale"] = l }
        if let t = info.timezone { dict["timezone"] = t }
        if let o = info.osVersion { dict["osVersion"] = o }
        if let appVersion = info.appVersion { dict["appVersion"] = appVersion }
        if let v = info.idfv { dict["idfv"] = v }
        if let a = info.idfa { dict["idfa"] = a }
        if let tk = info.adServicesToken { dict["adServicesToken"] = tk }
        if let attStatus = info.attStatus { dict["attStatus"] = attStatus.rawValue }
        result(dict)
      }

    case "getAdServicesToken":
      let token = AppSprintNative.getAdServicesToken()
      result(token as Any)

    case "requestTrackingAuthorization":
      Task {
        let authorized = await AppSprintNative.requestTrackingAuthorization()
        result(authorized)
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private static func numberValue(_ value: Any?) -> Double? {
    switch value {
    case let double as Double:
      return double
    case let float as Float:
      return Double(float)
    case let int as Int:
      return Double(int)
    case let number as NSNumber:
      return number.doubleValue
    case let string as String:
      return Double(string.trimmingCharacters(in: .whitespacesAndNewlines))
    default:
      return nil
    }
  }

  private static func googleAdsConsent(from value: Any?) -> GoogleAdsConsent? {
    guard let dict = value as? [String: Any],
          let status = googleAdsConsentStatus(from: dict["adUserData"]) else {
      return nil
    }
    return GoogleAdsConsent(adUserData: status)
  }

  private static func googleAdsConsentStatus(from value: Any?) -> GoogleAdsConsentStatus? {
    guard let raw = value as? String else {
      return nil
    }
    return GoogleAdsConsentStatus(rawValue: raw.trimmingCharacters(in: .whitespacesAndNewlines).uppercased())
  }

  private static func attributionToDictionary(_ attr: AttributionResult) -> [String: Any] {
    var dict: [String: Any] = [
      "isAttributed": attr.isAttributed,
      "source": attr.source,
      "confidence": attr.confidence,
    ]
    if let matchType = attr.matchType { dict["matchType"] = matchType }
    if let campaignName = attr.campaignName { dict["campaignName"] = campaignName }
    if let link = attr.link {
      dict["link"] = ["id": link.id, "name": link.name]
    }
    if let appleAds = attr.appleAds {
      var apple: [String: Any] = ["campaignId": appleAds.campaignId]
      if let orgId = appleAds.orgId { apple["orgId"] = orgId }
      if let adGroupId = appleAds.adGroupId { apple["adGroupId"] = adGroupId }
      if let keywordId = appleAds.keywordId { apple["keywordId"] = keywordId }
      if let adId = appleAds.adId { apple["adId"] = adId }
      if let country = appleAds.countryOrRegion { apple["countryOrRegion"] = country }
      if let claimType = appleAds.claimType { apple["claimType"] = claimType }
      if let clickDate = appleAds.clickDate { apple["clickDate"] = clickDate }
      if let impressionDate = appleAds.impressionDate { apple["impressionDate"] = impressionDate }
      if let conversion = appleAds.conversionType { apple["conversionType"] = conversion }
      if let supplyPlacement = appleAds.supplyPlacement { apple["supplyPlacement"] = supplyPlacement }
      dict["appleAds"] = apple
    }
    if let utmSource = attr.utmSource { dict["utmSource"] = utmSource }
    if let utmMedium = attr.utmMedium { dict["utmMedium"] = utmMedium }
    if let utmCampaign = attr.utmCampaign { dict["utmCampaign"] = utmCampaign }
    if let utmContent = attr.utmContent { dict["utmContent"] = utmContent }
    if let utmTerm = attr.utmTerm { dict["utmTerm"] = utmTerm }
    return dict
  }
}
